README.md
# laravel-cors

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.txt)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

This Laravel package provides support for enabling CORS requests in a Laravel application. If basic CORS support is needed, there are two other popular packages that are widely used ([barryvdh/laravel-cors](https://github.com/barryvdh/laravel-cors) and [spatie/laravel-cors](https://github.com/spatie/laravel-cors)).

This package attempts to solve a couple issues with the other packages. These issues are not show stoppers, so you may want to look at the other packages if these issues are not important to you.

#### 1. Multiple Configurations

While the other packages do support applying the CORS middleware to specific routes, they still only support one global configuration. So, for example, if you wanted one set of routes to be available to only one origin, and another set of routes to be available for a different origin, you can't create that configuration without making both sets of routes available to both origins.

This package allows specifying multiple configurations, which can be applied as needed per route.

#### 2. Error Handling

In the other packages, if an error occurred before the CORS middleware was applied, the CORS headers would not be added to the error response. In this case, instead of getting a response with the error, you would get an invalid CORS request error and not have access to the response.

This package works to ensure the CORS configuration is applied to the request, even in the case of an error.

#### 3. Preflight Issues

- Response Status Codes - Both packages return a 4xx status code for a rejected preflight request. The preflight request should still return a 2xx status code, it just needs to have the CORS response headers set (or not set) properly. This package returns a 204 No Body response for all preflight requests.

- Response Headers - Both packages don't set any headers for any rejected preflight request. This is appropriate for an invalid origin, but if the preflight is rejected due to the method or headers, the appropriate response headers should be added for the client to determine the proper rejection reason. Without any response headers, all rejection reasons will be due to an invalid origin. This package attempts to set the response headers smartly depending on the rejection scenario.

## Versions

This package is meant for Laravel 5.1 through Laravel 6.x. It does not yet support Lumen. This package uses middleware, and therefore does not support Laravel < 5.0. This package requires the `Kernel::hasMiddleware()` method, and therefore does not support Laravel 5.0.

## Install

Via Composer

``` bash
$ composer require shiftonelabs/laravel-cors
```

Once composer has been updated and the package has been installed, the service provider will need to be loaded.

#### Laravel 5.5+, 6.x (5.5, 5.6, 5.7, 5.8, 6.x)

This package uses auto package discovery. The service provider will automatically be registered.

#### Laravel 5.1 - 5.4

Open `config/app.php` and add following line to the providers array:

``` php
ShiftOneLabs\LaravelCors\LaravelCorsServiceProvider::class,
```

## Configuration

This package comes with a default configuration file that can be published in order to modify it.

``` bash
php artisan vendor:publish --provider="ShiftOneLabs\LaravelCors\LaravelCorsServiceProvider" --tag="config"
```

This will publish the config file to `config/cors.php`. Modify this config file to change the default profile and to add new custom profiles.

## Usage

#### Global Middleware

To apply a global CORS policy, add the `\ShiftOneLabs\LaravelCors\Http\Middleware\ApplyCorsPolicy` middleware to the global `$middleware` property in the `app/Http/Kernel.php` file. This global CORS policy will use whatever profile is defined as the default profile in the `cors.php` config file. By default, this is an open policy that will allow all CORS requests.

``` php
protected $middleware = [
    /* ... */,
    \ShiftOneLabs\LaravelCors\Http\Middleware\ApplyCorsPolicy::class,
]
```

#### Route Middleware

To apply a route-specific CORS policy, first add a middleware alias for the policy middleware to the `$routeMiddleware` property in the `app/Http/Kernel.php` file.

``` php
protected $routeMiddleware = [
    /* ... */,
    'cors.policy' => \ShiftOneLabs\LaravelCors\Http\Middleware\ApplyCorsPolicy::class,
]
```

Now you can add this middleware to any route or route group. Additionally, this route middleware can take a parameter to specify a CORS profile configuration defined in the `cors.php` config file.

Assume you add the following profile to the config file:

``` php
    'profiles' => [
        /* ... */,

        // add a new CORS profile named "app1"
        'app1' => [
            'allowedMethods' => ['*'],
            'allowedOrigins' => ['example.com'],
            'allowedOriginsPatterns' => [],
            'allowedHeaders' => ['*'],
            'exposedHeaders' => [],
            'maxAge' => 0,
            'supportsCredentials' => false,
        ],
    ],
```

This new CORS profile can be applied to any route or route group using route middleware:

``` php
Route::get('example-route', 'ExampleController@index')->middleware('cors.policy:app1');
```

This route will now only allow CORS requests from `example.com`. A CORS request from any other origin will be rejected.

#### Multiple Middleware

It is possible for a request to have to pass through multiple defined policies. For example, if you have a global CORS policy, a group middleware policy, and a route specific policy, the request must pass the requirements for all three policies.

If the request fails, the response headers will be built using the first policy that failed, starting with the least specific (global) to the most specific (route).

If the request is successful, the response headers will be built using the last policy that the request went through.

## Contributing

Contributions are welcome. Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email patrick@shiftonelabs.com instead of using the issue tracker.

## Alternatives

- [barryvdh/laravel-cors](https://github.com/barryvdh/laravel-cors): one of the most popular packages for adding CORS support to Laravel/Lumen applications. If you need simple support, go with this one.
- [spatie/laravel-cors](https://github.com/spatie/laravel-cors): another Laravel/Lumen CORS implementation from a well known package provider in the community. This package adds a little more flexibility than is available with the barryvdh/laravel-cors package.

## Credits

- [Patrick Carlo-Hickman][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.txt) for more information.

[ico-version]: https://img.shields.io/packagist/v/shiftonelabs/laravel-cors.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/shiftonelabs/laravel-cors/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/shiftonelabs/laravel-cors.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/shiftonelabs/laravel-cors.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/shiftonelabs/laravel-cors.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/shiftonelabs/laravel-cors
[link-travis]: https://travis-ci.org/shiftonelabs/laravel-cors
[link-scrutinizer]: https://scrutinizer-ci.com/g/shiftonelabs/laravel-cors/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/shiftonelabs/laravel-cors
[link-downloads]: https://packagist.org/packages/shiftonelabs/laravel-cors
[link-author]: https://github.com/patrickcarlohickman
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Initial release!

[Unreleased]: https://github.com/shiftonelabs/laravel-cors/compare/1.0.0...HEAD


tests/TestCase.php
<?php

namespace ShiftOneLabs\LaravelCors\Tests;

use Illuminate\Config\Repository;
use Illuminate\Foundation\Application;
use Illuminate\Support\Facades\Facade;
use Illuminate\Foundation\Testing\TestCase as BaseTestCase;

/**
 * @property \Illuminate\Foundation\Application $app
 */
class TestCase extends BaseTestCase
{
    use ReflectionTrait;

    protected $baseUrl = 'http://localhost';

    public function createApplication()
    {
        $app = new Application(dirname(__DIR__));
        $app->singleton(\Illuminate\Contracts\Http\Kernel::class, \ShiftOneLabs\LaravelCors\Tests\Fakes\Kernel::class);
        $app->singleton(\Illuminate\Contracts\Console\Kernel::class, \Illuminate\Foundation\Console\Kernel::class);
        $app->singleton(\Illuminate\Contracts\Debug\ExceptionHandler::class, \ShiftOneLabs\LaravelCors\Tests\Fakes\DebugExceptionHandler::class);

        // bootstrap: register config
        $app->instance('config', $config = new Repository([]));

        // bootstrap: register facades
        Facade::setFacadeApplication($app);

        // bootstrap: register providers
        $app->register(\ShiftOneLabs\LaravelCors\LaravelCorsServiceProvider::class);

        // bootstrap: boot providers
        $app->boot();

        return $app;
    }
}


tests/Fakes/Kernel.php
<?php

namespace ShiftOneLabs\LaravelCors\Tests\Fakes;

use Illuminate\Foundation\Http\Kernel as HttpKernel;

class Kernel extends HttpKernel
{
    /**
     * The bootstrap classes for the application.
     *
     * @var array
     */
    protected $bootstrappers = [
        //
    ];

    /**
     * The application's global HTTP middleware stack.
     *
     * These middleware are run during every request to your application.
     *
     * @var array
     */
    protected $middleware = [
        \Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode::class,
    ];

    /**
     * The application's route middleware.
     *
     * These middleware may be assigned to groups or used individually.
     *
     * @var array
     */
    protected $routeMiddleware = [
        'cors.policy' => \ShiftOneLabs\LaravelCors\Http\Middleware\ApplyCorsPolicy::class,
    ];
}


tests/CorsPolicyTest.php
<?php

namespace ShiftOneLabs\LaravelCors\Tests;

use Illuminate\Http\Request;
use Illuminate\Http\Response;
use PHPUnit_Framework_TestCase;
use ShiftOneLabs\LaravelCors\CorsPolicy;

class CorsPolicyTest extends PHPUnit_Framework_TestCase
{
    public function test_it_determines_if_a_request_is_allowed()
    {
        $request = new Request();
        $request->headers->set('Host', 'foo.com');
        $request->headers->set('Origin', 'http://bar.com');
        $request->setMethod('GET');

        // Request is allowed for global origins and global methods.
        $policy = new CorsPolicy(['allowedOrigins' => ['*'], 'allowedMethods' => ['*']]);
        $this->assertTrue($policy->isActualRequestAllowed($request));

        // Request is allowed for specific origins and specific methods.
        $policy = new CorsPolicy(['allowedOrigins' => ['http://bar.com', 'http://baz.com'], 'allowedMethods' => ['GET', 'POST']]);
        $this->assertTrue($policy->isActualRequestAllowed($request));

        // Request is not allowed for missing methods.
        $policy = new CorsPolicy(['allowedOrigins' => ['*'], 'allowedMethods' => ['POST']]);
        $this->assertFalse($policy->isActualRequestAllowed($request));

        // Request is not allowed for empty methods.
        $policy = new CorsPolicy(['allowedOrigins' => ['*'], 'allowedMethods' => []]);
        $this->assertFalse($policy->isActualRequestAllowed($request));

        // Request is not allowed for missing origins.
        $policy = new CorsPolicy(['allowedOrigins' => ['http://baz.com'], 'allowedMethods' => ['*']]);
        $this->assertFalse($policy->isActualRequestAllowed($request));

        // Request is not allowed for empty origins.
        $policy = new CorsPolicy(['allowedOrigins' => [], 'allowedMethods' => ['*']]);
        $this->assertFalse($policy->isActualRequestAllowed($request));
    }

    public function test_it_creates_origin_not_allowed_response()
    {
        $policy = new CorsPolicy();
        $response = $policy->createOriginNotAllowedResponse();

        $this->assertInstanceOf(Response::class, $response);
        $this->assertEquals(403, $response->getStatusCode());
    }

    public function test_it_creates_method_not_allowed_response()
    {
        $policy = new CorsPolicy();
        $response = $policy->createMethodNotAllowedResponse();

        $this->assertInstanceOf(Response::class, $response);
        $this->assertEquals(405, $response->getStatusCode());
    }

    public function test_it_creates_not_allowed_response()
    {
        $request = new Request();
        $request->headers->set('Host', 'foo.com');
        $request->headers->set('Origin', 'http://bar.com');
        $request->setMethod('GET');

        // Create 403 for origin not allowed.
        $policy = new CorsPolicy(['allowedOrigins' => [], 'allowedMethods' => []]);
        $response = $policy->createNotAllowedResponse($request);
        $this->assertInstanceOf(Response::class, $response);
        $this->assertEquals(403, $response->getStatusCode());

        // Create 405 for method not allowed.
        $policy = new CorsPolicy(['allowedOrigins' => ['*'], 'allowedMethods' => []]);
        $response = $policy->createNotAllowedResponse($request);
        $this->assertInstanceOf(Response::class, $response);
        $this->assertEquals(405, $response->getStatusCode());

        // Create 403 for any other reason.
        $policy = new CorsPolicy(['allowedOrigins' => ['*'], 'allowedMethods' => ['*']]);
        $response = $policy->createNotAllowedResponse($request);
        $this->assertInstanceOf(Response::class, $response);
        $this->assertEquals(403, $response->getStatusCode());
    }
}


src/CorsPolicy.php
<?php

namespace ShiftOneLabs\LaravelCors;

use Illuminate\Http\Request;
use Illuminate\Http\Response;
use Symfony\Component\HttpFoundation\Request as SymfonyRequest;

class CorsPolicy extends CorsService
{
    /**
     * Check the request origin and the request method to determine if the
     * request is allowed for this CORS policy.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    public function isActualRequestAllowed(SymfonyRequest $request)
    {
        return $this->isOriginAllowed($request) && $this->isMethodAllowed($request);
    }

    /**
     * Create the appropriate forbidden response based on the request.
     *
     * @param  \Illuminate\Http\Request  $request
     *
     * @return \Illuminate\Http\Response
     */
    public function createNotAllowedResponse(Request $request)
    {
        if (!$this->isOriginAllowed($request)) {
            return $this->createOriginNotAllowedResponse();
        }

        if (!$this->isMethodAllowed($request)) {
            return $this->createMethodNotAllowedResponse();
        }

        return $this->createResponse('Forbidden (cors).', 403);
    }

    /**
     * Create the response for an invalid request origin.
     *
     * @return \Illuminate\Http\Response
     */
    public function createOriginNotAllowedResponse()
    {
        return $this->createResponse('Origin not allowed.', 403);
    }

    /**
     * Create the response for an invalid request method.
     *
     * @return \Illuminate\Http\Response
     */
    public function createMethodNotAllowedResponse()
    {
        return $this->createResponse('Method not allowed.', 405);
    }

    /**
     * Create a response object.
     *
     * @param  string|null  $content
     * @param  int  $status
     * @param  array  $headers
     *
     * @return \Illuminate\Http\Response
     */
    protected function createResponse($content = '', $status = 200, $headers = [])
    {
        return new Response($content, $status, $headers);
    }
}


src/CorsPolicyManager.php
<?php

namespace ShiftOneLabs\LaravelCors;

use LogicException;
use Illuminate\Support\Arr;
use Illuminate\Support\Manager;

class CorsPolicyManager extends Manager
{
    /** @var array $config */
    protected $config;

    /**
     * Create a new manager instance.
     *
     * @param  \Illuminate\Foundation\Application  $app
     * @param  array  $config
     *
     * @return void
     */
    public function __construct($app, $config)
    {
        parent::__construct($app);

        $this->config = $config;
    }

    /**
     * Get the default driver name.
     *
     * @return string
     */
    public function getDefaultDriver()
    {
        return $this->config('default');
    }

    /**
     * Create a new Cors policy instance.
     *
     * @param  string  $profile
     *
     * @return \ShiftOneLabs\LaravelCors\CorsPolicy
     */
    public function make($profile = null)
    {
        return $this->driver($profile);
    }

    /**
     * Create a new driver instance.
     *
     * @param  string  $profile
     *
     * @return \ShiftOneLabs\LaravelCors\CorsPolicy
     *
     * @throws \LogicException
     */
    protected function createDriver($profile)
    {
        if (isset($this->customCreators[$profile])) {
            return $this->callCustomCreator($profile);
        }

        $config = $this->config('profiles.'.$profile);

        if (empty($config)) {
            throw new LogicException('CORS profile ['.$profile.'] not found.');
        }

        return new CorsPolicy($config);
    }

    /**
     * Get a value from the config array.
     *
     * @param  string  $key
     * @param  mixed|null  $default
     *
     * @return mixed
     */
    protected function config($key = null, $default = null)
    {
        if (empty($key)) {
            return $this->config;
        }

        return Arr::get($this->config, $key, $default);
    }
}


src/LaravelCorsServiceProvider.php
<?php

namespace ShiftOneLabs\LaravelCors;

use Illuminate\Contracts\Http\Kernel;
use Illuminate\Support\ServiceProvider;
use ShiftOneLabs\LaravelCors\Http\Middleware\CorsWrapper;

class LaravelCorsServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap any application services.
     *
     * @param  \Illuminate\Foundation\Http\Kernel  $kernel
     *
     * @return void
     */
    public function boot(Kernel $kernel)
    {
        $this->publishes([__DIR__.'/../config/cors.php' => config_path('cors.php')], 'config');

        $kernel->prependMiddleware(CorsWrapper::class);
    }

    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->mergeConfigFrom(__DIR__.'/../config/cors.php', 'cors');

        $this->app->singleton(CorsPolicyManager::class, function ($app) {
            return new CorsPolicyManager($app, $app['config']->get('cors'));
        });
    }
}


src/Http/Middleware/ApplyCorsPolicy.php
<?php

namespace ShiftOneLabs\LaravelCors\Http\Middleware;

use Closure;
use Illuminate\Routing\Router;
use Symfony\Component\HttpFoundation\Response;
use ShiftOneLabs\LaravelCors\CorsPolicyManager;

class ApplyCorsPolicy
{
    /** @var \ShiftOneLabs\LaravelCors\CorsPolicyManager $corsPolicyManager */
    protected $corsPolicyManager;

    /**
     * Create a new middleware instance.
     *
     * @param  \ShiftOneLabs\LaravelCors\CorsPolicyManager  $corsPolicyManager
     *
     * @return void
     */
    public function __construct(CorsPolicyManager $corsPolicyManager)
    {
        $this->corsPolicyManager = $corsPolicyManager;
    }

    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @param  string|null  $profile
     *
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function handle($request, Closure $next, $profile = null)
    {
        $cors = $this->corsPolicyManager->make($profile);

        if (!$cors->isCorsRequest($request)) {
            return $next($request);
        }

        if ($cors->isPreflightRequest($request)) {
            $preflightResponse = $cors->handlePreflightRequest($request);

            // Stop the call stack once a rejected preflight response is made.
            if ($cors->isPreflightRejected($preflightResponse)) {
                return $this->prepareResponse($preflightResponse);
            }

            // If this preflight was successful, move on to the next.
            $response = $next($request);

            // The destination core is plain text, not a response object. If we
            // hit the core, or the last middleware failed, return the current
            // preflight response. If we're on the way down the stack,
            // continue to return the last preflight response.
            return $this->prepareResponse($response instanceof Response && $response->isSuccessful() ? $response : $preflightResponse);
        }

        // Stop the call stack once a CORS request is rejected.
        if (!$cors->isActualRequestAllowed($request)) {
            return $this->prepareResponse($cors->createNotAllowedResponse($request));
        }

        $response = Router::toResponse($request, $next($request));

        // CORS policy already applied.
        if ($response->headers->has('X-S1L-CORS-HANDLED')) {
            return $this->prepareResponse($response);
        }

        return $this->prepareResponse($cors->addActualRequestHeaders($response, $request));
    }

    /**
     * Prepare the response that has been handled by the CORS middleware.
     *
     * @param  \Symfony\Component\HttpFoundation\Response  $response
     *
     * @return \Symfony\Component\HttpFoundation\Response
     */
    protected function prepareResponse(Response $response)
    {
        $response->headers->set('X-S1L-CORS-HANDLED', true);

        return $response;
    }
}


src/CorsService.php
<?php

namespace ShiftOneLabs\LaravelCors;

use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;

/**
 * Cors Service
 *
 * A lot of this code is derived from the asm89/stack-cors package. There were
 * a few things that needed tweaking, though, and the package's service class
 * uses private visibility, so I couldn't extend and override.
 *
 * @see https://github.com/asm89/stack-cors
 */
class CorsService
{
    /** @var array $options */
    protected $options;

    /**
     * Create a new service instance.
     *
     * @param  array  $options
     *
     * @return void
     */
    public function __construct(array $options = [])
    {
        $this->options = $this->normalizeOptions($options);
    }

    /**
     * Normalize the options into something more usable for the library.
     *
     * @param  array  $options
     *
     * @return array
     */
    protected function normalizeOptions(array $options = [])
    {
        $options += [
            'allowedOrigins' => [],
            'allowedOriginsPatterns' => [],
            'supportsCredentials' => false,
            'allowedHeaders' => [],
            'exposedHeaders' => [],
            'allowedMethods' => [],
            'maxAge' => 0,
        ];

        if (in_array('*', $options['allowedOrigins'])) {
            $options['allowedOrigins'] = true;
        }

        if (in_array('*', $options['allowedHeaders'])) {
            $options['allowedHeaders'] = true;
        } else {
            $options['allowedHeaders'] = array_map('strtolower', $options['allowedHeaders']);
        }

        if (in_array('*', $options['allowedMethods'])) {
            $options['allowedMethods'] = true;
        } else {
            $options['allowedMethods'] = array_map('strtoupper', $options['allowedMethods']);
        }

        return $options;
    }

    /**
     * Check if the request is a valid Cors request.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    public function isCorsRequest(Request $request)
    {
        return $request->headers->has('Origin') && $this->isCrossOrigin($request);
    }

    /**
     * Check if the request is a valid Cors preflight request.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    public function isPreflightRequest(Request $request)
    {
        return $this->isCorsRequest($request)
            && $request->getMethod() === 'OPTIONS'
            && $request->headers->has('Access-Control-Request-Method');
    }

    /**
     * Check the request origin to determine if the request is allowed.
     *
     * For the base CORS service, the request method is only checked during
     * preflight requests. If an invalid method is used for an actual
     * request, that should be handled at the application level.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    public function isActualRequestAllowed(Request $request)
    {
        return $this->isOriginAllowed($request);
    }

    /**
     * Add the valid Cors headers to the response.
     *
     * @param  \Symfony\Component\HttpFoundation\Response  $response
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function addActualRequestHeaders(Response $response, Request $request)
    {
        if (!$this->isActualRequestAllowed($request)) {
            return $response;
        }

        if (!$response->headers->has('Vary')) {
            $response->headers->set('Vary', 'Origin');
        } else {
            $response->headers->set('Vary', $response->headers->get('Vary') . ', Origin');
        }

        $response->headers->set('Access-Control-Allow-Origin', $request->headers->get('Origin'));

        if ($this->options['supportsCredentials']) {
            $response->headers->set('Access-Control-Allow-Credentials', 'true');
        }

        if ($this->options['exposedHeaders']) {
            $response->headers->set('Access-Control-Expose-Headers', implode(', ', $this->options['exposedHeaders']));
        }

        return $response;
    }

    /**
     * Get the full response for a Cors preflight request.
     *
     * This method deviates from the official preflight request algorithm. In the
     * official algorithm, if any of the origin, method, or headers is invalid,
     * no allow headers are added. So, if the origin is valid, but the header
     * is not supported, the allow origin header would not be added, and the
     * client would report an origin error, even though the origin is valid.
     *
     * This algorithm will only not add headers if the origin is not allowed (to
     * prevent leaking information). If the origin is allowed, then add all the
     * CORS response headers so the client can validate if the response is
     * valid and give the appropriate error message if not.
     *
     * The only exception to this algorithm is when the request method is not
     * allowed and it is a simple request method. For simple request methods,
     * clients do not validate the method against the allowed method header,
     * so we need to remove the allowed origin header to reject the request.
     *
     * @see https://www.w3.org/TR/cors/#resource-preflight-requests
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function handlePreflightRequest(Request $request)
    {
        // Preflight responses, even rejected ones, should return a 204
        // no body response.
        $response = $this->createResponse(null, 204);

        if (!$response->headers->has('Vary')) {
            $response->headers->set('Vary', 'Origin');
        } else {
            $response->headers->set('Vary', $response->headers->get('Vary') . ', Origin');
        }

        if (!$this->isOriginAllowed($request)) {
            return $response;
        }

        $response->headers->set('Access-Control-Allow-Origin', $request->headers->get('Origin'));

        // Clients ignore the Access-Control-Allow-Methods header for simple
        // request methods. In order to reject requests for simple methods
        // that aren't allowed, we disallow the origin.
        if (!$this->isMethodAllowed($request) && $this->isSimpleMethod($request)) {
            $response->headers->remove('Access-Control-Allow-Origin');
        }

        $allowMethods = $this->options['allowedMethods'] === true
            ? strtoupper($request->headers->get('Access-Control-Request-Method'))
            : implode(', ', $this->options['allowedMethods']);
        $response->headers->set('Access-Control-Allow-Methods', $allowMethods);

        $allowHeaders = $this->options['allowedHeaders'] === true
            ? strtoupper($request->headers->get('Access-Control-Request-Headers'))
            : implode(', ', $this->options['allowedHeaders']);
        $response->headers->set('Access-Control-Allow-Headers', $allowHeaders);

        if ($this->options['supportsCredentials']) {
            $response->headers->set('Access-Control-Allow-Credentials', 'true');
        }

        if ($this->options['maxAge']) {
            $response->headers->set('Access-Control-Max-Age', $this->options['maxAge']);
        }

        if ($this->isMethodAllowed($request) && $this->isHeadersAllowed($request)) {
            $response->headers->set('X-CORS-PREFLIGHT-SUCCESS', 'true');
        }

        return $response;
    }

    /**
     * Check if a given response is a successful preflight response.
     *
     * @param  \Symfony\Component\HttpFoundation\Response  $response
     *
     * @return bool
     */
    public function isPreflightSuccessful(Response $response)
    {
        // Only successful preflight responses will have this header.
        return $response->headers->has('X-CORS-PREFLIGHT-SUCCESS');
    }

    /**
     * Check if a given response is a rejected preflight response.
     *
     * @param  \Symfony\Component\HttpFoundation\Response  $response
     *
     * @return bool
     */
    public function isPreflightRejected(Response $response)
    {
        return !$this->isPreflightSuccessful($response);
    }

    /**
     * Check if the request is actually cross origin (the origin isn't the same
     * as the request host).
     *
     * @param \Symfony\Component\HttpFoundation\Request $request
     *
     * @return bool
     */
    protected function isCrossOrigin(Request $request)
    {
        return $request->headers->get('Origin') !== $request->getSchemeAndHttpHost();
    }

    /**
     * Check if the request origin is allowed by the CORS configuration.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    protected function isOriginAllowed(Request $request)
    {
        if ($this->options['allowedOrigins'] === true) {
            return true;
        }

        $origin = $request->headers->get('Origin');

        if (in_array($origin, $this->options['allowedOrigins'])) {
            return true;
        }

        foreach ($this->options['allowedOriginsPatterns'] as $pattern) {
            if (preg_match($pattern, $origin)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Check if the request method is allowed by the CORS configuration.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    protected function isMethodAllowed(Request $request)
    {
        if ($this->options['allowedMethods'] === true) {
            return true;
        }

        $method = $this->getActualRequestMethod($request);

        return in_array($method, $this->options['allowedMethods']);
    }

    /**
     * Check if the request headers are allowed by the CORS configuration.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    protected function isHeadersAllowed(Request $request)
    {
        if ($this->options['allowedHeaders'] === true || !$request->headers->has('Access-Control-Request-Headers')) {
            return true;
        }

        $headers = strtolower($request->headers->get('Access-Control-Request-Headers'));
        $headers = array_filter(array_map('trim', explode(',', $headers)));

        return empty(array_diff($headers, $this->options['allowedHeaders']));
    }

    /**
     * Check if this a simple request method.
     *
     * @see https://www.w3.org/TR/cors/#simple-method
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return bool
     */
    protected function isSimpleMethod(Request $request)
    {
        $method = $this->getActualRequestMethod($request);

        return in_array($method, ['GET', 'HEAD', 'POST']);
    }

    /**
     * Get the request method for the actual request, even if the current
     * request is a preflight request.
     *
     * @param  \Symfony\Component\HttpFoundation\Request  $request
     *
     * @return string
     */
    protected function getActualRequestMethod(Request $request)
    {
        return $this->isPreflightRequest($request)
            ? strtoupper($request->headers->get('Access-Control-Request-Method'))
            : $request->getMethod();
    }

    /**
     * Create a response object.
     *
     * @param  string|null  $content
     * @param  int  $status
     * @param  array  $headers
     *
     * @return \Symfony\Component\HttpFoundation\Response
     */
    protected function createResponse($content = '', $status = 200, $headers = [])
    {
        return new Response($content, $status, $headers);
    }
}
